### PR TITLE
Allow override of bundler version or resolve from lock file

### PR DESCRIPTION
**Pull request by inverse**
Allow override of the default provided bundler from the image or resolve from `Gemfile.lock` if one exists.

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -8,6 +8,13 @@ if [ -n "$INPUT_PRE_BUILD_COMMANDS" ]; then
   eval "$INPUT_PRE_BUILD_COMMANDS"
 fi 
 
+echo "INPUT_BUNDLER_VERSON:${INPUT_BUNDLER_VERSON}"
```
Comment (inverse) [inline entrypoint.sh:11]: For some reason this is coming back as empty when testing with:

```
      - uses: inverse/jekyll-action@override-bundler
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          bundler_version: '2.2.31'
          ```

Comment (helaili) [reply entrypoint.sh:11]: Replace `VERSON` with `VERSION` maybe?

Comment (inverse) [reply entrypoint.sh]: How did I miss that 🙈

**Comment by inverse** [comment]
@helaili anything you would change on this?

**Comment by helaili** [comment]
Sorry @inverse, time is a scarce resource. 
Would you mind adding a test case similar to [this one](https://github.com/helaili/jekyll-action/blob/master/.github/workflows/test.yml#L72-L117)?

**Comment by inverse** [comment]
@helaili I added two cases cases where it would do what you outlined - I think it's correct?

I also added asserting the bundler version matches either explicitly or from the lockfile.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -69,6 +71,115 @@ jobs:
...
+  jekyll_bundler_from_lock_test: 
+    runs-on: ubuntu-latest
+    needs: jekyll_bundler_publish
+    steps:
+    - uses: actions/checkout@v2
+    - uses: ./.github/actions/test
+      with: 
+        token: ${{ secrets.GITHUB_TOKEN }}
+        spec: cypress/integration/jekyll_src/**/*
+        suffix: "jekyll_src"
```
Comment (helaili) [inline .github/workflows/test.yml:181]:
```suggestion
        spec: cypress/integration/jekyll_bundler_lock/**/*
        suffix: "jekyll_bundler_lock"
```

**Review by helaili — CHANGES_REQUESTED** [review]
Thanks for working on this @inverse. I have only reviewed the test generation so far but I'd like it to be properly run before I review the rest of this PR.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -1,7 +1,9 @@
 name: Testing the GitHub Pages publication
 
 on:
-  push
+  push:
+    branches: [master]
+  pull_request: ~
```
[the remark this answers is not in the document]
Comment (inverse) [reply .github/workflows/test.yml:6]: I think it's just for the default and is same as just having `pull_request:` We can revert it

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -69,6 +71,115 @@ jobs:
...
+    - name: Jekyll src run
+      id: run
+      uses: ./
+      with:
+        jekyll_build_options: "--config ../sample_site/_config.yml,../sample_site/_config_jekyll_src.yml"
+        jekyll_src: sample_site
+        target_branch: jekyll_src_test_pages_branch
+        jekyll_env: production
+      env:
+        JEKYLL_PAT: ${{ secrets.GITHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (inverse) [reply .github/workflows/test.yml:150]: Do you mean like this

```

diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 7baa4fd..5d9f837 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -146,8 +146,8 @@ jobs:
         jekyll_src: sample_site
         target_branch: jekyll_bundler_lock_test_pages_branch
         jekyll_env: production
+        token: ${{ secrets.GITHUB_TOKEN }}
       env:
-        JEKYLL_PAT: ${{ secrets.GITHUB_TOKEN }}
         JEKYLL_DEBUG: true
     - id: extract-bundler-version
     ```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -69,6 +71,115 @@ jobs:
...
+    - name: Jekyll bundler run
+      id: run
+      uses: ./
+      with:
+        jekyll_build_options: "--config ../sample_site/_config.yml,../sample_site/_config_jekyll_bundler.yml"
+        jekyll_src: sample_site
+        target_branch: jekyll_bundler_test_pages_branch
+        jekyll_env: production
+        bundler_version: 2.3.11
+        github-token: ${{secrets.JEKYLL_PAT}}
```
Comment (helaili) [inline .github/workflows/test.yml:95]: My bad, sorry I got confused with the `github-script` action.
```suggestion
        token: ${{secrets.JEKYLL_PAT}} 
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -69,6 +71,115 @@ jobs:
...
+        github-token: ${{secrets.JEKYLL_PAT}} 
+      env:
+        JEKYLL_DEBUG: true
+    - id: extract-bundler-version
+      run: echo ::set-output name=value::$(bundler -v | cut -c 16- | xargs)
+    - uses: nick-fields/assert-action@v1
+      with:
+        expected: 2.3.11
+        actual: ${{ steps.extract-bundler-version.outputs.value }}
+        comparison: exact
```
Comment (helaili) [inline .github/workflows/test.yml:104]: This step is failing. Please check https://github.com/helaili/jekyll-action/runs/6257218854?check_suite_focus=true.

**Review by helaili — CHANGES_REQUESTED** [review]
Tests are not passing

File: .github/workflows/test.yml (hunk 6 above)
Comment (inverse) [reply .github/workflows/test.yml:104]: Just pushed something that should fix it, forgot to quote the set-output

**Comment by helaili** [comment]
Tests are still failing. Trying to debug it, I've seen [this line](https://github.com/helaili/jekyll-action/runs/6272309590?check_suite_focus=true#step:5:13). 

Thing is `bundler` is being installed within the container, and the `echo "::set-output name=value::$(bundler -v | cut -c 16- | xargs)"` command is being called within the next step, so outside of the container where `bundler` is not installed or could be a totally different one. 

You need to run the `set-output` command from `entrypoint.sh` and referred to the value as coming from the `run` step. This step should actually be renamed for clarity.

**Comment by inverse** [comment]
Ahh you're totall right :see_no_evil: 

> You need to run the set-output command from entrypoint.sh and referred to the value as coming from the run step. This step should actually be renamed for clarity.

What do you mean by that? I don't follow?

I  pushed a change which instead does the following:

- Adds the bundler version and exposes it on the page
- Modifies the cypress tests for the two additional flows to assert the bundler version

image

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -69,6 +71,101 @@ jobs:
...
+        jekyll_src: sample_site
+        target_branch: jekyll_bundler_lock_test_pages_branch
+        jekyll_env: production
+        token: ${{ secrets.GITHUB_TOKEN }}
+      env:
+        JEKYLL_DEBUG: true
+        
+  jekyll_bundler_from_lock_publish: 
+    runs-on: ubuntu-latest
+    needs: [basic_test, jekyll_bundler_from_lock_run]
```
Comment (helaili) [inline .github/workflows/test.yml:148]:
```suggestion
    needs: [jekyll_bundler_test, jekyll_bundler_from_lock_run]
```

**Review by helaili — CHANGES_REQUESTED** [review]
I think it's getting closed. We can unfortunately publish and test in sequence as there is only one published site at a time, so the publish needs to happen after the last test and then we run the test.

**Comment by inverse** [comment]
@helaili good catch! how does the graph look now? https://github.com/helaili/jekyll-action/actions/runs/2296309955

**Comment by inverse** [comment]
@helaili I think I know why...

So this test case is when we force install a newer version of bundler than what the application is bundled with. Yet the system has `2.1.4` so it bundles with that version as it's in the lockfile :/

Not sure what a good test would be for this case? 


I could remove that part and we just install the bundler version from the lockfile which probably covers the majority of cases...

**Comment by helaili** [comment]
I'm removing the Gemfile.lock file during the `jekyll_bundler_run` job so that it is generated with the right `Bundler` version. I'm also now relying on the `bundle -v` command as opposed to `bundler -v` to figure out the version number as `bundler is not in the path in this specific case. 

Thanks for this contribution @inverse

**Comment by inverse** [comment]
@helaili That's for taking the time to explain :+1:

**Pull request merged**